### PR TITLE
win32: Support loading libsodium dynamically

### PR DESCRIPTION
**Pull request by k-takata**
Only MS-Windows for now.

### CODE REVIEW

File: src/crypt.c [hunk 1]
```diff
@@ -990,4 +1092,24 @@ crypt_sodium_buffer_decode(
...
 }
 
+# if defined(FEAT_SODIUM) || defined(PROTO)
+    int
+crypt_sodium_munlock(void *const addr, const size_t len)
+{
+# if DYNAMIC_SODIUM
+    // FIXME:
+    // The free_buf_options() function in buffer.c may call this without
+    // calling crypt_sodium_init().
```
Comment (k-takata) [inline src/crypt.c]: I found that `sodium_munlock()` may be called from `free_buf_options()` without calling `crypt_sodium_init()`.

Steps to reproduce:
1. Start vim: `gvim -u DEFAULTS`
2. Add some text: `ahello<Esc>`
3. Enable libsodium: `:set cm=xchacha20`
4. Set encrypt key: `:X` (e.g. "hello")
5. Write to a file: `:w hello.txt`
6. `sodium_munlock()` is called without `crypt_sodium_init()`.

@chrisbra 
This is unintended, right?

Comment (chrisbra) [reply src/crypt.c]: yes, I don't think this should happen

Comment (k-takata) [reply src/crypt.c]: The issue has been fixed: https://github.com/vim/vim/pull/9554/commits/7cb7f9c09310d451896677ebe676c9f822fa616d

File: src/crypt.c [hunk 2]
```diff
@@ -159,6 +159,108 @@ typedef struct {
...
+    }
+
+    for (i = 0; sodium_funcname_table[i].ptr; ++i)
+    {
+	if ((*sodium_funcname_table[i].ptr = GetProcAddress(hsodium,
+			sodium_funcname_table[i].name)) == NULL)
+	{
+	    FreeLibrary(hsodium);
+	    hsodium = NULL;
+	    // TODO: Show error message.
```
Comment (k-takata) [inline src/crypt.c:257]: I haven't added error messages here, but I think PR is ready now.

**Comment by chrisbra** [comment]
Hi @k-takata,
I think we should add something to the documentation on how to make sodium-encryption work on Windows. Because I think for a user who wants to make use of it, it is unclear that they need to download libsodium.dll and place it in the path (or vims installation directory). What do you think?

**Comment by k-takata** [comment]
Agree.
And, I think it would be nice to include libsodium.dll in the vim-win32-installer packages. Your patch is needed now: https://github.com/vim/vim-win32-installer/issues/220#issuecomment-879672746

**Pull request closed without merging**